#Tarea Fogler P 10-9c
##Reacción para la producción del i-octano a partir de i-octeno e hidrógneo

#### i-octeno + hidrógeno => i-octano
$A+B\rightarrow C$

#### Partimos de proponer el mecanismo, basandonos en las ecuaciones de LHHW:


1.   $A+S \rightleftharpoons AS$
2.   $B+S \rightleftharpoons BS$
3.   $AB+BS\rightleftharpoons CS+S$
4.   $CS\rightleftharpoons C+S$

####Donde se asume que las velocidades de adsorción y de reacción son nulas, es decir:

*   $C_{A}C_{v}-\frac{C_{AS}}{K_{A}}=0$
*   $C_{B}C_{v}-\frac{C_{BS}}{K_{B}}=0$
*   $C_{AS}C_{BS}-\frac{C_{CS}C_{v}}{K_{S}}=0$

####Y que $-r_{A}$, por lo tanto es:
$-r_{A}=\frac{k_{S}}{C_{t}}\left ( C_{AS}C_{BS} -\frac{C_{CS}C_{v}}{K_{S}} \right )$

####para, que haciendo las sustituciones pertinentes en _$-r_{A}$_ dar la siguiente expresión:
$-r_{A} = \frac{k\left ( P_{A}P_{B}- \frac{P_{C}}{K_{eq}} \right )}{\left [ 1 +K_{A}P_{A} + K_{B}P_{B} + K_{C}P_{C}\right ]^{2}}$




In [1]:
import pandas as pd

####Creamos el dataset a partir de los datos que viene en el ejercicio

In [2]:
df = pd.DataFrame({'experimento': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], 'velocidad': [0.0362, 0.0239, 0.0390, 0.0351, 0.0114, 0.0534, 0.0280, 0.0033, 0.0380, 0.0090, 0.0127, 0.0566], 'hidrogeno': [1,1,3,1,1,10,1,1,2,1,0.6,5], 'octeno':[1,1,1,3,1,1,10,1,2,1,0.6,5],"octano": [0,1,1,1,3,0,0,10,2,4,0.6,5]})

In [3]:
print(df)

    experimento  velocidad  hidrogeno  octeno  octano
0             1     0.0362        1.0     1.0     0.0
1             2     0.0239        1.0     1.0     1.0
2             3     0.0390        3.0     1.0     1.0
3             4     0.0351        1.0     3.0     1.0
4             5     0.0114        1.0     1.0     3.0
5             6     0.0534       10.0     1.0     0.0
6             7     0.0280        1.0    10.0     0.0
7             8     0.0033        1.0     1.0    10.0
8             9     0.0380        2.0     2.0     2.0
9            10     0.0090        1.0     1.0     4.0
10           11     0.0127        0.6     0.6     0.6
11           12     0.0566        5.0     5.0     5.0


####La ecuación que sale del posible mecanismo de reacción es: 
$-r_{A} = \frac{k\left ( P_{A}P_{B}- \frac{P_{C}}{K_{eq}} \right )}{\left [ 1 +K_{A}P_{A} + K_{B}P_{B} + K_{C}P_{C}\right ]^{2}}$
####Y su linearización es:
$\left ( \frac{{P_{A}P_{B}}}{-r_{A}} \right )^{1/2} = \frac{1}{k^{1/2}} + \frac{K_{A}P_{A}}{k^{1/2}} +\frac{K_{B}P_{B}}{k^{1/2}}+ \frac{K_{C}P_{C}}{k^{1/2}}$

##### Donde A es el _i-octeno_, B es el _hidrogeno_ y C es el _i-octano_

####Para linealizar obtenemos la columna llamada _**CaCb/-ra**_ 

In [4]:

 df['CaCb/-ra'] = df.apply(lambda row: (row.octeno * row.hidrogeno / row.velocidad)**(0.5), axis=1)

In [5]:
print(df)

    experimento  velocidad  hidrogeno  octeno  octano   CaCb/-ra
0             1     0.0362        1.0     1.0     0.0   5.255883
1             2     0.0239        1.0     1.0     1.0   6.468462
2             3     0.0390        3.0     1.0     1.0   8.770580
3             4     0.0351        1.0     3.0     1.0   9.245003
4             5     0.0114        1.0     1.0     3.0   9.365858
5             6     0.0534       10.0     1.0     0.0  13.684514
6             7     0.0280        1.0    10.0     0.0  18.898224
7             8     0.0033        1.0     1.0    10.0  17.407766
8             9     0.0380        2.0     2.0     2.0  10.259784
9            10     0.0090        1.0     1.0     4.0  10.540926
10           11     0.0127        0.6     0.6     0.6   5.324139
11           12     0.0566        5.0     5.0     5.0  21.016568


####Para hacer la regresión lineal necesitamos especificar nuestras variables independientes, es decir, la concentración de hidrógeno, la concentración del **octeno** y la concentración del **octano**

In [6]:
X_multiple = df[['hidrogeno', 'octeno','octano']]
print(X_multiple)

    hidrogeno  octeno  octano
0         1.0     1.0     0.0
1         1.0     1.0     1.0
2         3.0     1.0     1.0
3         1.0     3.0     1.0
4         1.0     1.0     3.0
5        10.0     1.0     0.0
6         1.0    10.0     0.0
7         1.0     1.0    10.0
8         2.0     2.0     2.0
9         1.0     1.0     4.0
10        0.6     0.6     0.6
11        5.0     5.0     5.0


####Y definimos el valor de nuestra variable dependiente *y*, es decir, _**CaCb/-ra**_

In [7]:
Y_multiple = df[['CaCb/-ra']]
print(Y_multiple)

     CaCb/-ra
0    5.255883
1    6.468462
2    8.770580
3    9.245003
4    9.365858
5   13.684514
6   18.898224
7   17.407766
8   10.259784
9   10.540926
10   5.324139
11  21.016568


####Importamos las librerías para hacer la regresión lineal múltiple

In [8]:
from sklearn.model_selection import train_test_split
from sklearn import linear_model

####Dividimos nuestros datos en datos de entrenamiento _**D_train**_ y datos de prueba _**D_test**_

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X_multiple, Y_multiple, test_size=0.35)

####Guardamos el modelo en una variable llamada  _regresion_lineal_multiple_

In [38]:
regresion_lineal_multiple = linear_model.LinearRegression()

####Entrenamos el modelo, es decir, hacemos la regresión lineal


In [39]:
regresion_lineal_multiple.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

####Ya hecha la regresión lineal obtenemos los coeficientes de la regresión donde el primer valor es para el hidrógeno (B), el segundo valor para el i-octeno (A) y el tercer valor para el i-octano (C)

In [40]:
regresion_lineal_multiple.coef_

array([[0.90088291, 1.47903169, 1.20982094]])

####Obtenemos el valor en la intersección 

In [41]:
regresion_lineal_multiple.intercept_

array([3.15786522])

####Por último obtenemos el coeficiente de R² para ver qué tan buena es nuestra predicción

In [42]:
regresion_lineal_multiple.score(X_train,y_train)

0.9992204794696232

####Por lo tanto la ecuación de la regresión múltiple nos queda de la siguiente manera: 
$y = 3.17617368 + 1.47772837 P_{A} + 0.9034848 P_{B} + 1.20982094 P_{C}$
####Con un coeficiente de determinación r² = 0.9992204794696232

##Conclusiones:
###Se propuso un mecanismo que es consistente con los datos experimentales y
###Se econtró la velocidad de reaccion evaluando todos los parámetros